jsonschema/_types.py

```diff
@@ -0,0 +1,206 @@
+import numbers
+
+import attr
+import pyrsistent
+
+from jsonschema.compat import str_types, int_types, iteritems
+from jsonschema.exceptions import UndefinedTypeCheck
+
+
+def is_array(instance):
+    return isinstance(instance, list)
+
+
+def is_bool(instance):
+    return isinstance(instance, bool)
+
+
+def is_integer(instance):
+    # bool inherits from int, so ensure bools aren't reported as ints
+    if isinstance(instance, bool):
+        return False
+    return isinstance(instance, int_types)
+
+
+def is_null(instance):
+    return instance is None
+
+
+def is_number(instance):
+    # bool inherits from int, so ensure bools aren't reported as ints
+    if isinstance(instance, bool):
+        return False
+    return isinstance(instance, numbers.Number)
+
+
+def is_object(instance):
+    return isinstance(instance, dict)
+
+
+def is_string(instance):
+    return isinstance(instance, str_types)
+
+
+def is_any(instance):
+    return True
+
+
+@attr.s(frozen=True)
+class TypeChecker(object):
+    """
+    A ``type`` property checker.
+
+    A :class:`TypeChecker` performs type checking for an instance of
+    :class:`Validator`. Type checks to perform are set using
+    :meth:`TypeChecker.redefine` or :meth:`TypeChecker.redefine_many` and
+    removed via :meth:`TypeChecker.remove` or
+    :meth:`TypeChecker.remove_many`. Each of these return a new
+    :class:`TypeChecker` object.
+
+    Arguments:
+
+        type_checkers (pyrsistent.pmap):
+
+            It is recommend to set type checkers through
+            :meth:`TypeChecker.redefine` or :meth:`TypeChecker.redefine_many`
+    """
+    _type_checkers = attr.ib(default=pyrsistent.pmap({}))
+
+    def is_type(self, instance, type):
+        """
+        Check if the instance is of the appropriate type.
+
+        Arguments:
+
+            instance (any primitive type, i.e. str, number, bool):
+
+                The instance to check
+
+            type (str):
+
+                The name of the type that is expected.
+
+        Returns:
+
+            bool: Whether it conformed.
+
+
+        Raises:
+
+            :exc:`jsonschema.exceptions.UndefinedTypeCheck`:
+                if type is unknown to this object.
+        """
+        try:
+            return self._type_checkers[type](instance)
+        except KeyError:
+            raise UndefinedTypeCheck
```

Comment:
Should include the name of the type here (and ideally add that to the test that checks this branch)